debug print improvements

biddisco opened the pull request:
This PR contains a load of debug printing improvements from several branches I'm working on, in particular latest libfabric changes. Pushing this to see if it causes any compilation problems on dashboards, as I will need it merged before other branches.

hkaiser commented on libs/debugging/include/hpx/debugging/print.hpp:
```diff
@@ -225,71 +307,201 @@ namespace hpx { namespace debug {
         }
 
+#else
+        // C++14 version
+        // helper function to print a tuple of any size
+        template <typename TupleType, std::size_t N>
+        struct tuple_printer
+        {
+            static void print(std::ostream& os, const TupleType& t)
+            {
+                tuple_printer<TupleType, N - 1>::print(os, t);
+                os << " " << std::get<N - 1>(t);
+            }
+        };
+
+        template <typename TupleType>
+        struct tuple_printer<TupleType, 1>
+        {
+            static void print(std::ostream& os, const TupleType& t)
+            {
+                os << std::get<0>(t);
+            }
+        };
+
+        template <class... Args>
+        void tuple_print(std::ostream& os, const std::tuple<Args...>& t)
+        {
+            tuple_printer<decltype(t), sizeof...(Args)>::print(os, t);
+        }
```
FWIW, something like the following would avoid the recursive expansion:
```suggestion
        // C++14 version
        // helper function to print a tuple of any size
        namespace detail {
            template <typename Tuple, std::size_t... Is>
            void tuple_print(std::ostream& os, Tuple const& t, 
                 std::index_sequence<Is...>)
            {
                using expander = int[];
                (void) expander{0, (void(os << ' ' << std::get<Is>(t)), 0)...};
            }
       }
        
        template <typename... Args> 
        void tuple_print( 
            std::ostream& os, std::tuple<Args...>const& t) 
        { 
            detail::tuple_print(os, t, std::make_index_sequence<sizeof...(Args)>{});
        }
```
```

biddisco replied:
I'll try this out, thanks

hkaiser commented on libs/debugging/include/hpx/debugging/print.hpp:
```diff
@@ -225,71 +307,201 @@ namespace hpx { namespace debug {
+            }
+
+            int guess_rank() const {
+#ifdef DEBUGGING_PRINT_LINUX
```
WHat's the rationale of limiting this to `LINUX`?

biddisco replied:
I didn't want to add and env var checking things that might not compile on windows without testing them. I'm guessing that I'm the only person using this debug print stuff, so I didn't think anyone would notice.

hkaiser replied:
The code you wrapped should be working fine on other platforms as long as you use:
```
#ifdef __APPLE__
#include <crt_externs.h>
#define environ (*_NSGetEnviron())
#elif defined(HPX_WINDOWS)
#define environ _environ
#else
extern char** environ;
#endif
```

hkaiser commented on libs/threading_base/include/hpx/threading_base/print.hpp:
```diff
@@ -126,31 +125,48 @@ namespace hpx { namespace debug {
+#ifdef HPX_HAVE_CXX17_FOLD_EXPRESSIONS
         template <typename... Args>
-        void display(const char* prefix, Args... args)
+        void display(const char* prefix, std::forward<Args>(args)... args)
```
This doesn't look right.

biddisco replied:
It isn't right. not sure how that crept in, but I've removed it.

biddisco commented:
> Normally we use east-const notation in our code base... I'd appreciate it if you adapted to this. Thanks!

Is this something that clang-format can do fo us? I am not accustomed to writing consts on the right and I don't want to change all this by hand.

hkaiser commented:
> > Normally we use east-const notation in our code base... I'd appreciate it if you adapted to this. Thanks!
> 
> Is this something that clang-format can do fo us? I am not accustomed to writing consts on the right and I don't want to change all this by hand.

The clang-format people  say it's not the job of their tool to do that (clang-format never changes the sequence of tokens). They think this is something clang-tidy could provide. I'm not sure if somebody has actually implemented this there, however.

biddisco commented:
> > The clang-format people say it's not the job of their tool to do that (clang-format never changes the sequence of tokens). They think this is something clang-tidy could provide. I'm not sure if somebody has actually implemented this there, however.

There is work on-going in clang-format to add support for an east-west const fixer
https://reviews.llvm.org/D69764
I knocked up a quick regex search to find all the places where we are using west const instead of east const - it's a lot. I am going to push a cleanup patch to fix the cxx-14 variadic print (thanks, I've used your patch, plus some other improvments) and then a separate one with east const fixes for the debug printer.

hkaiser commented:
> There is work on-going in clang-format to add support for an east-west const fixer
> https://reviews.llvm.org/D69764

Cool, I was not aware of that.

> I knocked up a quick regex search to find all the places where we are using west const instead of east const - it's a lot. I am going to push a cleanup patch to fix the cxx-14 variadic print (thanks, I've used your patch, plus some other improvments) and then a separate one with east const fixes for the debug printer.

Thanks, much appreciated.

hkaiser commented on libs/debugging/include/hpx/debugging/print.hpp:
```diff
@@ -23,11 +25,14 @@
 #define DEBUGGING_PRINT_LINUX
 #endif
 
+#undef HPX_HAVE_CXX17_FOLD_EXPRESSIONS
```
What's the rationale of always disabling the use of fold expressions here?

biddisco commented:
I made a bunch of cleanups and fixed some old stuff that wasn't updated. @hkaiser was correct in saying that debug arguments that are function returns are evaluated even in debug<false> paths, so I added a lazy `eval` wrapper (had to use a macro unfortunately) and added a unit test that checks that the functions/features are behaving as tested. Any other tweaks can wait for a new PR or the next release.

hkaiser commented on libs/debugging/tests/unit/print.cpp:
```diff
@@ -0,0 +1,129 @@
+    return ++counter;
+}
+
+#define AB_LAZY(expr, debug) \
```
Is this is a typo? The code below uses `DB_LAZY`.

biddisco replied:
Sorry. Left over from an experiment. Removed

hkaiser commented on libs/debugging/include/hpx/debugging/print.hpp:
```diff
@@ -61,6 +64,11 @@
 
+// Used to wrap function call parameters to prevent evaluation
+// when debugging is disabled
+#define DP_LAZY(Expr, printer) \
```
Please prefix the macro name with `HPX_` to avoid unwanted name-clashes with macro names defined in other places.

biddisco replied:
done

biddisco commented:
Add the environ check as suggested. I actually plan on removing that stuff and replacing it with something that uses the batch environment checks that already exist - but I will leave that until after the mpi branch is finished as I am actually using this in the mpi testing at the moment and it can wait as it is not urgent.

hkaiser commented:
Let's get this is asap.

biddisco commented:
After several rounds of fixing the windows/mac builds and sorting out some silly warnings etc - this now looks ok on dashboards - the fails being returned from tests are unrelated.

hkaiser commented on libs/debugging/include/hpx/debugging/print.hpp:
```diff
@@ -23,9 +25,17 @@
+#include <unistd.h>
+#define environ (*_NSGetEnviron())
+#elif defined(HPX_WINDOWS)
+#include <winsock2.h>
```
Why do we need to `#include <winsock2.h>` now?

hkaiser replied:
The symbol `char** _environ` is defined in `#include <stdlib.h>` if this is what you were looking for.

biddisco replied:
The CI windows build failed, so I replaced winsock with winsock2 and now it works. I haven't used windows in a while and don't really remember the details - the call to `gethostname` was the problem. It seems to be compiling without errors or warnings on all our test machines now.
